server/src/main/java/org/apache/druid/segment/realtime/plumber/RealtimePlumber.java

```diff
@@ -454,7 +454,7 @@ public void doRun()
               DataSegment segment = dataSegmentPusher.push(
                   mergedFile,
                   sink.getSegment().withDimensions(IndexMerger.getMergedDimensionsFromQueryableIndexes(indexes)),
-                  false
+                  true
```

Comment:
Why'd you change this to `true`? It will involve creating extra segments on deep storage if people have replicas with their realtime plumbers.

I thought about https://github.com/apache/incubator-druid/pull/7558 while reviewing this line, because I was about to say on a review here that BaseAppenderatorDriver is able to safely delete the unpublished replicas' pushed segments. I was going to say that unfortunately the technique is not applicable here, because it relies on seeing a publish failure, but without transactional metadata checks _all_ replicas will have their publishes succeed. (RealtimePlumber doesn't do transactional metadata checks.)

Reply:
I see, have changed it back to false.

Reply:
writing to tmp and then doing `rename` provided atomic write ... now if there are 3 replicas , trying to push concurrently ... who wins and is it guaranteed that end result will not be corrupted ?